### PR TITLE
ceph-volume: check unlocked devices for VGs

### PR DESCRIPTION
**Pull request by jan--f**
After deactivating a volume group a disk is no longer rejected (in inventory) as its no longer locked. This change adds a check for volume groups that have space available (>=5GB) and only then marks such a disk as available.
    
Fixes: https://tracker.ceph.com/issues/43400
Signed-off-by: Jan Fajerski <jfajerski@suse.com>

Note that this is based on #32235, which is ready to merge. Only the last commit in this series contains the fix mentioned.

### CODE REVIEW

File: src/ceph-volume/ceph_volume/util/device.py [hunk 1]
```diff
@@ -394,6 +388,9 @@ def _check_reject_reasons(self):
             rejected.append('Insufficient space (<5GB)')
         if self.is_ceph_disk_member:
             rejected.append("Used by ceph-disk")
+        available_vgs = [vg for vg in self.vgs if vg.free >= 5368709120]
```
Comment (leseb) [inline src/ceph-volume/ceph_volume/util/device.py]: I think this only works when not using the upcoming raw mode for preparing OSDs (https://github.com/ceph/ceph/pull/32095).
I think it's a bit of a corner case here, I'd think that if a VG has >=5GB of free space, then c-v was presented LVs to create OSDs from that VG but actually never created the PV/VG, right?
Reporting available by scanner the VG only seems quite conservative with the LVM approach.

I'd hope that there is another way to reject that device but currently the raw mode would certainly fail on such a device. But again I don't think we will see that often.

Thoughts?

Thanks for addressing the issue rapidly.

Comment (jan--f) [reply src/ceph-volume/ceph_volume/util/device.py]: > I think this only works when not using the upcoming raw mode for preparing OSDs (#32095).

Right, the raw mode will need additional code. We need to query ceph-bluestore-tool for this.

> I think it's a bit of a corner case here, I'd think that if a VG has >=5GB of free space, then c-v was presented LVs to create OSDs from that VG but actually never created the PV/VG, right?
> Reporting available by scanner the VG only seems quite conservative with the LVM approach.
> 
> I'd hope that there is another way to reject that device but currently the raw mode would certainly fail on such a device. But again I don't think we will see that often.
> 
> Thoughts?

I started writing some code to give user a say in what is considered available. I think this is the way to go. E.g. by default inventory assumes `--lvm` and it'll use the current set of availability tests. Passing `--raw` would use a different set. 

> 
> Thanks for addressing the issue rapidly.

No worries at all, I was going to say sorry it took so long ;)

Comment (leseb) [reply src/ceph-volume/ceph_volume/util/device.py]: I think inventory should be available to discover all sets without the need to pass an args like `--lvm` or `--raw`. The inventory should be really generic.
Can we have a small test if the device is not an LVM member to calculate the free space on the disk?

Thanks.

Comment (jan--f) [reply src/ceph-volume/ceph_volume/util/device.py]: I agree generic would be ideal. But consider the case where you already have a ceph lv and a vg with 50g free space. That is a valid device for c-v lvm, not so for c-v raw.

Maybe adding a inventory subcommand to the various modes is the way to go. then one can call c-v [lvm|raw] inventory for specific tests or c-v inventory for a conjunction of the various subcommands?


> Can we have a small test if the device is not an LVM member to calculate the free space on the disk?

For the inventory output? Iiuc we test this already and return unavailable if there is less then 5GB free.

**Comment by jan--f** [comment]
Closed in favour of #32634

**Pull request closed without merging**